[#43323] Nextcloud: Validate :host to provide a Secure Context (HTTPS or loopback/localhost)

fraber opened the pull request:
https://community.openproject.org/work_packages/43323

Specification for **Secure Context**
https://w3c.github.io/webappsec-secure-contexts/

It also changes how we validate `redirect_uri` of `::Doorkeeper::Application`, AKA "OAuth applications". Now it allows more types of loopback addresses and not only 'localhost'.

fraber committed:
Add validation for file storages: Host attribute needs to provide a secure context

Kharonus commented on modules/storages/app/validator/nextcloud_secure_host_validator.rb:
```diff
@@ -0,0 +1,44 @@
+    end
+
+    return if uri.scheme == 'https' # https is always safe
+    return if ['localhost', '127.0.0.1'].include?(uri.host) # localhost is always safe
```
Not sure if this matching suffices. Usually a host is considered secure for `127.0.0.0/8` or `::1/128`, and of course for the host names `localhost` and `localhost.`, or if they end with `.localhost` or `.localhost.`.

wielinde replied:
Adding to it: If you make the list allowed hosts a constant then that list gets easily patchable in case somebody needs to add another exception, let's say by building a OpenProject plugin.

wielinde commented on modules/storages/app/validator/nextcloud_secure_host_validator.rb:
```diff
@@ -0,0 +1,44 @@
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+class NextcloudSecureHostValidator < ActiveModel::EachValidator
```
I believe this validator does not hold anything Nextcloud specific. Or am I wrong?
```suggestion
class SecureContextURIValidator < ActiveModel::EachValidator
```

wielinde commented on modules/storages/app/validator/nextcloud_secure_host_validator.rb:
```diff
@@ -0,0 +1,44 @@
+      return
+    end
+
+    return if uri.scheme == 'https' # https is always safe
```
```suggestion
    return if uri.scheme == 'https' # https is always safe
    return if uri.host == 'localhost' # Simple localhost
    return if uri.host =~ /\.localhost\.?$/ # i.e. 'foo.localhost' or 'foo.localhost.'
    return if (IPAddr.new(uri.host).loopback? rescue false) # Constructor throws an exception for non IP addresses. Those are invalid. And if the host is an IP address then we can check if it as loopback
```

wielinde commented on modules/storages/app/validator/nextcloud_secure_host_validator.rb:
```diff
@@ -0,0 +1,44 @@
+    end
+
+    return if uri.scheme == 'https' # https is always safe
+    return if ['localhost', '127.0.0.1'].include?(uri.host) # localhost is always safe
```
```suggestion
    return if uri.host == 'localhost' # Simple localhost
    return if uri.host =~ /\.localhost\.?$/ # i.e. 'foo.localhost' or 'foo.localhost.'
    return if (IPAddr.new(uri.host).loopback? rescue false) # Constructor throws an exception for non IP addresses. Those are invalid. And if the host is an IP address then we can check if it as loopback
```

wielinde commented on modules/storages/app/validator/nextcloud_secure_host_validator.rb:
```diff
@@ -0,0 +1,44 @@
+# See COPYRIGHT and LICENSE files for more details.
+#++
+class NextcloudSecureHostValidator < ActiveModel::EachValidator
+  # A "secure" Nextcloud host is either a host via https or
```
@fraber Please correct the comment here and add the link to the specification of Secure Context https://w3c.github.io/webappsec-secure-contexts/

fraber commented on app/validators/secure_context_uri_validator.rb:
```diff
@@ -0,0 +1,67 @@
+  def check_invalid_uri(uri)
+    return false if uri.scheme == 'https' # https is always safe
+    return false if uri.host == 'localhost' # Simple localhost
+    return false if uri.host =~ /\.localhost\.?$/ # i.e. 'foo.localhost' or 'foo.localhost.'
```
This regex looks like redundant (with the IPAddr.new below), but it isn't, IPAddr.new doesn't get these cases.

wielinde replied:
Correct, `IPAddr` only handles IP addresses and not URIs.

wielinde commented on spec/validator/secure_context_uri_validator_spec.rb:
```diff
@@ -0,0 +1,98 @@
+          subject.host = uri
+          subject.validate
+          expect(subject.errors).to include(:host)
+          # expect(subject.errors.include?(:host)).to be_truthy
```
Comment line...

wielinde commented on spec/validator/secure_context_uri_validator_spec.rb:
```diff
@@ -0,0 +1,98 @@
+  subject do
+    Class.new do
+      include ActiveModel::Validations
+      def self.model_name
```
I have no idea why this is necessary. Any explanation?

fraber replied:
ActiveModel::Errors needs the name of the class in order to format the error messages used below. With this model_name the error messages won't appear.
The alternative would be to use a non-anonymous class.

wielinde commented on spec/validator/secure_context_uri_validator_spec.rb:
```diff
@@ -0,0 +1,98 @@
+    Class.new do
+      include ActiveModel::Validations
+      def self.model_name
+        ActiveModel::Name.new(self, nil, "temp")
```
What does this line do?

fraber replied:
see above, the line belongs to def self.model_name. It produces the class name "temp", which is appropriate :-)

wielinde replied:
Ah, I got off railed as class names in Ruby start with a capital letter. So we better rename it to
```suggestion
        ActiveModel::Name.new(self, nil, "TempModel")
```

wielinde replied:
I finally renamed it to `ValidatedModel`

wielinde commented on spec/validator/secure_context_uri_validator_spec.rb:
```diff
@@ -0,0 +1,98 @@
+  end
+
+  describe 'secure URI' do
+    ['https://www.example.com', 'http://localhost', 'http://.localhost', 'http://foo.localhost.'].each do |uri|
```
I believe 'http://foo.localhost' is missing in the list

wielinde commented on spec/validator/secure_context_uri_validator_spec.rb:
```diff
@@ -0,0 +1,98 @@
+
+  describe 'empty URI' do
+    ['', ' ', nil].each do |uri|
+      describe "when URI is ''#{uri}''" do
```
Can't we use propper double quotes:
`describe "when URI is \"#{uri}\"' do`

wielinde commented on app/validators/secure_context_uri_validator.rb:
```diff
@@ -25,20 +25,43 @@
-    return if uri.scheme == 'https' # https is always safe
-    return if ['localhost', '127.0.0.1'].include?(uri.host) # localhost is always safe
+    if check_invalid_uri(uri)
+      contract.errors.add(attribute, :host_not_https_or_localhost)
```
IMHO a better name for `:host_not_https_or_localhost` would be `:uri_not_secure_context`

wielinde commented on app/validators/secure_context_uri_validator.rb:
```diff
@@ -25,20 +25,43 @@
+# Please see https://w3c.github.io/webappsec-secure-contexts/
+# for a definition of "secure contexts".
+# Basically, a host has to have either a HTTPS scheme or be
+# localhost to be secure.
```
.... to provide a secure context.

wielinde commented on spec/validator/secure_context_uri_validator_spec.rb:
```diff
@@ -0,0 +1,98 @@
+
+  describe 'invalid URI' do
+    ['nope', '.', 'httppp://192.168.0.1', 'http://192.168', 'http://<>ample.com'].each do |uri|
+      describe "when URI is ''#{uri}''" do
```
again, can't we simply use escaped double quotes? `\"` instead of two single ticks?

wielinde commented on spec/validator/secure_context_uri_validator_spec.rb:
```diff
@@ -0,0 +1,98 @@
+    end
+  end
+
+  describe 'secure URI' do
```
`describe` is the wrong term here. It is a `context`.

wielinde commented on spec/validator/secure_context_uri_validator_spec.rb:
```diff
@@ -0,0 +1,98 @@
+    end.new
+  end
+
+  describe 'empty URI' do
```
`describe` is the wrong term here. It is a `context`.

wielinde commented on spec/validator/secure_context_uri_validator_spec.rb:
```diff
@@ -0,0 +1,98 @@
+
+  describe 'secure URI' do
+    ['https://www.example.com', 'http://localhost', 'http://.localhost', 'http://foo.localhost.'].each do |uri|
+      describe "when URI is #{uri}" do
```
`describe` is the wrong term here. It is a `context`.

wielinde commented on spec/validator/secure_context_uri_validator_spec.rb:
```diff
@@ -0,0 +1,98 @@
+  # There are alternative forms of apparently valid IPV6 localhost
+  # notation not supported by URI.parse in SecureContextUriValidator,
+  # such as "::1" or "0001:0000:0000:0000:0000:0000:0000:0000".
+  describe 'secure IPV6 URI' do
```
`describe` is the wrong term here. It is a `context`.

wielinde commented on spec/validator/secure_context_uri_validator_spec.rb:
```diff
@@ -0,0 +1,98 @@
+    end
+  end
+
+  # There are alternative forms of apparently valid IPV6 localhost
```
The comment here is not correct. `http://::1` is not a valid URI. I bet `http://0001:0000:0000:0000:0000:0000:0000:0000` neither.

wielinde replied:
What is `0001:0000:0000:0000:0000:0000:0000:0000` actually? It doesn't seem to be a loopback address.

wielinde commented on spec/validator/secure_context_uri_validator_spec.rb:
```diff
@@ -0,0 +1,98 @@
+    ['', ' ', nil].each do |uri|
+      describe "when URI is ''#{uri}''" do
+        it "adds an 'could_not_parse' error" do
+          subject.host = uri
```
These two lines are the same for every example. They can get extracted to a `before` block

fraber replied:
Unfortunately this doesn't work out because it's not lazy evaluated. We'd have to reformulate using let(:...).

wielinde replied:
Ok, what stops us?

wielinde replied:
I changed it so that every generated context simply sets the `host` and nothing else.

wielinde commented on app/validators/secure_context_uri_validator.rb:
```diff
@@ -25,20 +25,43 @@
+      uri = nil
+    end
+
+    if uri.nil? || uri.host.nil?
```
`uri&.host.present?` would do the trick. Not sure if that would improve readability by much.

wielinde replied:
I hope my latest changes make it easier to read.

wielinde reviewed (changes requested):
Cool, that you have a explicit spec for the validator that tests all nitty gritty details. This reduces nicely what needs to be tested in contracts using the validator.

Before merging it needs a couple of improvements, though. Please checkout my comments.

wielinde commented:
I reworked the spec a bit to also cover those cases where a URI was parsable but not providing a Secure Context.

wielinde committed:
Use same Secure Context validation for OAuth applications

Currently redirect_uri field of OAuth applications (Doorkeeper
Applications) did only check for 'localhost', which is not
complete. Other loopback URLs are also valid, such as
'http://127.0.0.1'.

Now, it is coherent with the allowed hosts of ::Storages::Storage

wielinde committed:
Remove redundant error flash. I don't a reason, why it is needed.

Kharonus commented on modules/storages/spec/contracts/storages/storages/shared_contract_examples.rb:
```diff
@@ -116,6 +116,18 @@
+        let(:storage_host) { 'http://nc.openproject.com' }
+
+        include_examples 'contract is invalid', host: :url_not_secure_context
+      end
```
Can we add a positive and negative example with ipv6? This has two reasons: 

- we check if it works, also for future changes on the validator
- there are a minority of users, who check tests for usage examples. So they could check, how an ipv6 addr can be provided (it must have the brackets, mustn't it?)

UPDATE:
just saw, that we have those test cases directly for the validator. this suffices in my opinion

Kharonus reviewed (commented):
code wise this looks good to me, I will make a test now

Kharonus reviewed (changes requested):
If I now run this test, I get a rails error: `NoMethodError`

specifically this is about the call in line 48 in `secure_context_uri_validator`

can you verfiy that? Can't figure, how this could be related to my setup?

wielinde commented:
@Kharonus It was a bug in my code. That spec now passes locally. Let's see what CI is saying, now.

Kharonus commented:
@wielinde : seems to have fixed the bug, but there is an expectation in that unit test, that still fails.
